src/types/notimplemented.rs

```diff
@@ -15,24 +15,26 @@ impl PyNotImplemented {
     }
 }
 
-impl<'v> PyTryFrom<'v> for PyNotImplemented {
-    fn try_from<V: Into<&'v PyAny>>(value: V) -> Result<&'v Self, crate::PyDowncastError<'v>> {
-        let value: &PyAny = value.into();
-        if unsafe { ffi::Py_NotImplemented() == value.as_ptr() } {
-            return unsafe { Ok(value.downcast_unchecked()) };
-        }
-        Err(PyDowncastError::new(value, "NotImplementedType"))
+unsafe impl PyTypeInfo for PyNotImplemented {
+    const NAME: &'static str = "NotImplementedType";
+
+    const MODULE: Option<&'static str> = None;
+
+    type AsRefTarget = PyNotImplemented;
+
+    fn type_object_raw(_py: Python<'_>) -> *mut ffi::PyTypeObject {
+        unsafe { ffi::Py_TYPE(ffi::Py_NotImplemented()) }
     }
 
-    fn try_from_exact<V: Into<&'v PyAny>>(
-        value: V,
-    ) -> Result<&'v Self, crate::PyDowncastError<'v>> {
-        value.into().downcast()
+    #[inline]
+    fn is_type_of(object: &PyAny) -> bool {
+        // NoneType is not usable as a base type
+        Self::is_exact_type_of(object)
     }
 
-    unsafe fn try_from_unchecked<V: Into<&'v PyAny>>(value: V) -> &'v Self {
-        let ptr = value.into() as *const _ as *const PyNotImplemented;
-        &*ptr
+    #[inline]
+    fn is_exact_type_of(object: &PyAny) -> bool {
+        object.is(PyNotImplemented::get(object.py()))
```

Comment:
I think we should add `PyAny::is_not_implemented` and either use these methods for all three implementations or but the three implementations here and rewrite the methods in terms of them.

Reply:
I quite like `is_none()` but personally I think `is_not_implemented()` and `is_ellipsis()` are needed rarely enough that they're not worth pulling their weight.

There are also the three getter methods `py.None()`, `py.NotImplemented()` and `py.Ellipsis()`. Really these are just a slightly nicer way to write `PyNone::get(py)`. (They return `PyObject` but really they should return `&PyNone` etc in my opinion, I'll open a separate PR for that in a moment.)

What do you think about deprecating `obj.is_ellipsis()` in favour of `obj.is(py.Ellipsis())`?

Reply:
I decided to open #3578 specifically to change the return types. I'll come back to this PR afterwards and do the `is_ellipsis()` deprecation here as well as tidy up these implementations.

Reply:
Rebased to implement all three by `object.is(Self::get(py))`, and also deprecate `is_ellipsis`.